upload_vehicle_data.py

```diff
@@ -0,0 +1,740 @@
+        delete_remote_temp_directory(
+            logger, remote_user, remote_ip, remote_directory
+        )
+
+
+def main(config, debug):
+    """Automate the upload of rosbags."""
+    logger = setup_logging(debug)
+    remote_user = config['remote_user']
+    remote_ip = config['remote_ip']
+    base_remote_directory = config['remote_directory']
+    cloud_upload_directory = config['cloud_upload_directory']
+    directory_depth = config.get(
+        'directory_depth', 1
+    )  # Default to 1 for flat structure
+    logger.info('Starting rosbag upload process.')
+
+    # Measure bandwidth once at the start
+    bandwidth_mbps = measure_bandwidth(logger, remote_ip, remote_user)
+    if bandwidth_mbps is None:
+        logger.error('Could not measure bandwidth. Exiting.')
+        return
+    # Get all subdirectories in the base remote directory
+    subdirectories = list_remote_directories(
+        logger, remote_user, remote_ip, base_remote_directory, directory_depth
+    )
+    logger.debug(f'Rosbags subdirectories found: {len(subdirectories)}')
+    total_rosbags = 0
+    total_size_gb = 0.0
+    total_estimated_time = 0.0
+    file_sizes_dict = {}
+
+    # Compute total estimated time for all subdirectories
+    for subdirectory in subdirectories:
+        # Get the list of rosbags from the remote machine
+        rosbag_list = get_remote_rosbags_list(
+            logger, remote_user, remote_ip, subdirectory
+        )
+
+        # Get the size of each rosbag
+        rosbag_sizes = [
+            get_remote_file_size(logger, remote_user, remote_ip, rosbag)
+            for rosbag in rosbag_list
+        ]
+        file_sizes_dict[subdirectory] = rosbag_sizes
+        total_rosbags += len(rosbag_list)
+        total_size_gb += sum(rosbag_sizes)
+        total_estimated_time += get_estimated_upload_time(
+            sum(rosbag_sizes), bandwidth_mbps, rosbag_sizes
+        )
+
+    if total_estimated_time < 1:
+        estimated_time_str = f'{total_estimated_time * 60:.2f} minutes'
+    else:
+        estimated_time_str = f'{total_estimated_time:.2f} hours'
+
+    logger.info(
+        f'Found {total_rosbags} files to upload '
+        f'with total size {total_size_gb:.2f} GB.'
+    )
+    logger.info(
+        f'Estimated total time (including compression) for '
+        f'all subdirectories is: {estimated_time_str}.'
+    )
+    confirm = input('Do you want to proceed to upload? (yes/no): ')
+
+    if confirm.lower() != 'yes':
+        logger.info('Upload aborted by user.')
+        return
+
+    logger.info('User confirmed upload. Beginning processing of directories.')
+
+    # Process each subdirectory
+    for subdirectory in subdirectories:
+        process_directory(
+            logger,
+            remote_user,
+            remote_ip,
+            subdirectory,
+            cloud_upload_directory,
+            config,
+            base_remote_directory,
+            bandwidth_mbps,
+            file_sizes_dict,
+        )
+
+

```

Comment:
We are missing the final log after processing all of the subdirectories. This log should mention how many files were successfully uploaded. We will need find a way to keep track of those numbers

````python
logger.info(f'Uploading finished. {total_uploaded_files}/{total_files} files were successfully uploaded.')

````

Reply:
Changed